packages/drivers/odsp-driver/src/opsCaching.ts

```diff
@@ -0,0 +1,167 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+// ISequencedDocumentMessage
+export interface IMessage {
+    sequenceNumber: number;
+}
+
+type CacheEntry = (IMessage | undefined)[];
+
+interface IBatch {
+    currentBatchSize: number;
+    batchData: CacheEntry;
+    dirty: boolean;
+}
+
+export interface ICache {
+  write(batchNumber: string, data: string): Promise<void>;
+  read(batchNumber: string): Promise<string | undefined>;
+}
+
+export class OpsCache {
+  private readonly batches: Map<number, null | IBatch> = new Map();
+  private timer: ReturnType<typeof setTimeout> | undefined;
+
+  constructor(
+    startingSequenceNumber: number,
+    private readonly cache: ICache,
+    private readonly batchSize: number,
+    private readonly timerGranularity,
+    private totalOpsToCache,
+  ) {
+        /** initial batch is a special case because it will never be full - all ops prior (inclusive) to
+         *  startingSequenceNumber are never going to show up (undefined)
+         */
+        const currentBatchSize = this.getPositionInBatchArray(startingSequenceNumber) + 1; // inclusive
+        if (currentBatchSize !== this.batchSize) {
+            this.batches.set(this.getBatchNumber(startingSequenceNumber), {
+                currentBatchSize,
+                batchData : this.initializeNewBatchDataArray(),
+                dirty: false,
```

Comment:
Should not the dirty be true here?

Reply:
No, false is correct value as there are no ops to flush yet.
I'm adding comment for dirty - it tells if batch has any ops that were not flushed to cache yet